cachi2/core/package_managers/cargo/main.py

```diff
@@ -28,13 +64,65 @@ def fetch_cargo_source(request: Request) -> RequestOutput:
     return RequestOutput.from_obj_list(components, environment_variables, project_files)
 
 
-def _resolve_cargo_package(package_dir: RootedPath, output_dir: RootedPath) -> list[Component]:
+def _extract_package_info(path_to_toml: Path) -> dict:
+    # 'value' unwraps the underlying dict and that makes mypy happy (it complains about
+    # mismatching type otherwise despite parsed document having the necessary interface).
+    return tomlkit.parse(path_to_toml.read_text()).value["package"]
+
+
+def _resolve_main_package(package_dir: RootedPath) -> dict:
+    try:
+        return _extract_package_info(package_dir.path / "Cargo.toml")
+    # We'll get here in the case of virtual workspaces. A real-world example is
+    # https://github.com/rwf2/Rocket/tree/master
+    # In this case there is no package name per se, but there still is a metapackage.
+    # We'll use directory name as a fallback in this case.
+    # Version won't make much sense here: this is a meta-package, the state of a
+    # repository will be captured in VCS_URL, and individual components will be
+    # versioned.
+    except KeyError:
+        return {
+            "name": package_dir.path.stem,
+            "version": None,
+        }
+
+
+def _verify_lockfile_is_present_or_fail(package_dir: RootedPath) -> None:
+    # Most packages will be locked, however metapackages (i.e. those, which
+    # contain just a workspace and could even lack a name) could arrive without
+    # a lock file. A user could try and fix this by explicitly locking the
+    # package first.
+    if not (package_dir.path / "Cargo.lock").exists():
+        raise PackageRejected(
+            f"{package_dir.path} is not locked",
+            solution="Please lock it first by running 'cargo generate-lockfile",
+        )
```

Comment:
Hold on, so what is the common practice with Cargo workspaces wrt to lockfiles? Is there supposed to be just a single lockfile on the repo root level or is there supposed to be one for each workspace package since there is one `Cargo.toml` per workspace member?

Reply:
Okay, thanks for clarifying, then it's reasonable for us to ask the user to do it.